### ISSUE TITLE
Apply MAP-Elites to Picbreeder

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#615]
Idea from paper "Are Quality Diversity Algorithms Better at Generating Stepping Stones than Objective-based Search?"

Apply MAP-Elites to Picbreeder. Should be fairly easy since my Innovation Engine code already does this. See how this works.

However, the next step is to implement the better versions of MAP-Elites based on CMA-ES that Amy is currently working with

**Comment by schrum2** [comment]
Note: The paper mentioned above seems to sort of already succeed at recreating "the skull" from Picbreeder. Still, would be fun to see if I can do the same.

The specific bin dimensions they use for MAP-Elites are "complexity" which is just the number of CPPN links, and "novelty" which is measured using an Autoencoder. I would definitely try something simpler first, but the autoencoder idea is interesting. 

> "we introduce a generic notion of novelty. Periodically all of the images in the map are compiled and used to train an autoencoder [3]; when a new image is produced we compress and reconstruct the image with the autoencoder, examining the pixel-by-pixel error of the reconstruction. Images similar to those which were used to train the autoencoder will be reconstructed well, those different poorly. The reconstruction error is used as generic approximation of “novelty” or “interestingness”

The idea of measuring novelty with an autoencoder makes sense, but I don't think it makes sense as a MAP-Elites bin dimension ... unless I'm misunderstanding something. Basically, I don't see why you would want to save into than archive an image that was uniquely un-novel.

Furthermore, if I want to use an Autoencoder, I'll have to use DL4J ... that should be possible, and could even be interesting, but let's see what I can do without it first.

**Comment by schrum2** [comment]
We'll definitely have to talk about this before you attempt it. 

We can't do the autoencoder approach right away.

For our first shot at this, let's make the MAP-Elites bin dimensions be the number of CPPN neurons and the number of CPPN links. This is probably too simple, but it is a starting point. The fitness would be the pixel-by-pixel correspondence to a particular target image.

**Comment by schrum2** [comment]
One more note. There is a class `PictureInnovationTask` that is already similar to what you will be doing here. You'll basically start by copying `edu.southwestern.tasks.innovationengines.PictureInnovationTask` to `edu.southwestern.tasks.innovationengines.TargetPictureTask`

All mention of ImageNet will be removed, and you will replace the `behaviorVector` with completely difference calculations, as well as make other changes.

Still, this is an advanced task to address after doing other, simpler tasks

**Comment by annazwicker** [comment]
![image](https://user-images.githubusercontent.com/82399865/118285557-32c89680-b497-11eb-950f-b06229542a69.png)

**Comment by schrum2** [comment]
The authors of the short 2-page paper do not fully clarify what the fitness/quality measure in their experiments is. They just indicate that is it some sort of comparison between the evolved image and the target image. However, it seems very likely that they based their experiments off of previous work by Brian Woolley, so the exact fitness is probably the one defined in this paper:
https://eplex.cs.ucf.edu/papers/woolley_gecco11.pdf
Specifically, look at equation (3) and preceding associated equations on page 4.

However, this fitness seems to only be for black/white (grayscale) images, so we will need to change our code to produce grayscale images before working with color.

**Comment by schrum2** [comment]
There is a Parameter called `blackAndWhitePicbreeder` that makes black and white images evolve. Try setting this in the main method of `PictureTargetTask` and see what happens

**Comment by annazwicker** [comment]
Done: Added the equations from the Woolley paper, updated GraphicsUtil to include grayscale images.
Error calculation seems higher than it should be, but not necessarily wrong.  Something to look into later.  The skull image included approximately 25 neurons and 53 links.  Currently waiting on the program to finish running using the skull as the target image.

**Comment by schrum2** [comment]
Since 53 is way more than 35, we should set the max num links for our binning scheme to about 70 before running another test. We should also make sure that the list of activation functions we are using matches what is in the original paper. I found this quote:

"As in Picbreeder, the available CPPN activation functions were sine, cosine, Gaussian, identity, and sigmoid." We should restrict the activation functions we use to contain only these (might also look at the original skull genome to see exactly which versions of each function it uses ... I know that my code supports multiple types of Gaussian and sigmoid).

Another potentially relevant quote from the paper is this paragraph, but I'm not sure exactly what it means:
"To compare two images and calculate fitness, each is defined by a feature set that includes the grayscale value at each pixel location (at 128 × 128 pixels) and the gradient between adjacent pixel values"
We're using the grayscale value at each pixel, but we are not using the gradient between adjacent pixel values ... in fact, I'm not exactly sure what they mean by this. We'll try to make due without it though, for now at least.

**Comment by annazwicker** [comment]
A new plot showing the maximum individual fitness has been created (accompanying this is a new method in StatisticsUtilities and some slight changes in MAPElites).  This graph currently appears to show the correct data with the correct scale.  The variable MAX_NUM_LINKS has also been increased since the skull target image had approximately 53 links (it has now been increased to 70).  However, the images produced are still pretty far off from the skull target image.  The program is still currently running and will be left to run overnight.

**Comment by annazwicker** [comment]
![0 54838Neurons 17 links 61 103260](https://user-images.githubusercontent.com/82399865/118687244-6888ba80-b7ca-11eb-9e52-ffb5dfcb0a52.jpg)
![0 69179Neurons 35 links 69 566176](https://user-images.githubusercontent.com/82399865/118686182-65d99580-b7c9-11eb-973b-6a6dd7881130.jpg)
![0 69881Neurons 35 links 63 747971](https://user-images.githubusercontent.com/82399865/118686402-94577080-b7c9-11eb-9ae9-05ad94ce9294.jpg) 

The last image has the highest individual fitness but isn't incredibly interesting or close to the target image of the skull.  The first image however, has a lower fitness but has a lot more contrast similar to the target image.  Perhaps something is wrong with our fitness calculation (taken directly from the Woolley paper).  All of our activation functions match the Woolley paper as well.  However, the maximum number of links is 70 (this was changed since the skull image itself had about 53).  Increasing this number resulted in more interesting images, but is relatively high.

**Comment by schrum2** [comment]
I think that ultimately, we will need to test out a variety of different fitness functions to make this task work, so at the point in the code where fitness is calculated, we'll actually need an if/else if/else structure that checks if various command line parameters are true. Basically, we'll have a boolean command line parameter for each fitness function we want to try. 

However, before this makes sense to do, we have to have at least one other specific fitness function we want to try. Perhaps we could use the loss function computed by the autoencoder in the DL4J code #635

**Comment by schrum2** [comment]
Ok, Adam (author of the stepping stones paper) responded with some helpful details.

The fitness they used was RMSE (root mean squared error). Specifically, in his dissertation (https://hal.univ-lorraine.fr/tel-02964666/document) he says:
"Fitness is simply the root mean square error of the difference in pixel
intensities of the produced and target image"

Although I believe a calculation of this should be readily available in DL4J, let's start by implementing it ourselves. The formula is widely available in various places, such as this one:
https://www.statisticshowto.com/probability-and-statistics/regression-analysis/rmse-root-mean-square-error/

Specifically, make a method or methods in the `StatisticsUtilities` class for computing the RMSE.

Then, have boolean command line parameters (as described above) for enabling different fitness functions. Have a parameter "picTargetWoolleyFitness" for the fitness function we are currently using, and have "picTargetRMSE" use the new RMSE you will make.

**Comment by schrum2** [comment]
Although we have not successfully evolved the Skull image yet, all remaining work for this issue has been moved to other issues, namely #642 and #647 . However, we can run an evolution experiment that tries to evolve a target picture using MAP Elites and one of two possible fitness functions, so I'm closing this issue.
